File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.java
class Pair{
    int a,b;
    Pair(int a, int b){
        this.a=a;
        this.b=b;
    }
}

class Solution {
    public int minimumDeletions(String s) {

        Map<Integer,Pair>map=new HashMap<>();
        
        int ct1=0,ct2=0;
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='a')ct1++;
            else ct2++;
            map.put(i,new Pair(ct1,ct2));
        }

        int ans=Integer.MAX_VALUE;

        for(int i=0;i<s.length();i++){
                Pair p1=map.get(i);
                Pair p2=map.get(s.length()-1);    

            if(s.charAt(i)=='a'){
                ans=Math.min(ans,p1.b+p2.a-p1.a);
            }
            else{
                ans=Math.min(ans,p1.b-1+p2.a-p1.a);
                
            }
        }

        return ans;


    }
}

File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        int n=s.size();
        vector<int>count(n);
        int ct=0;
        for(int i=n-1;i>=0;i--){
            if(s[i]=='a')ct++;
            count[i]=ct;
        }

        int ans=INT_MAX;
        ct=0;
        for(int i=0;i<n;i++){
            if(s[i]=='a')continue;
            if(s[i]=='b' && count[i]==0)break;
            ans=min(ans,ct+count[i]);
            ct++;
        }
        return min(ct,ans);
    }
};